test/Microsoft.Azure.Devices.Edge.Test.Common/Runtime.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Test.Common
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Test.Common.Config;
+    using Microsoft.Azure.Devices.Edge.Util;
+
+    using Registries = System.Collections.Generic.IEnumerable<(string address, string username, string password)>;
+
+    public class Runtime
+    {
+        readonly string agentImage;
+        readonly string deviceId;
+        readonly string hubImage;
+        readonly IotHub iotHub;
+        readonly bool optimizeForPerformance;
+        readonly Option<Uri> proxy;
+        readonly Registries registries;
+
+        public Runtime(string deviceId, string agentImage, string hubImage, Option<Uri> proxy, Registries registries, bool optimizeForPerformance, IotHub iotHub)
+        {
+            this.agentImage = agentImage;
+            this.deviceId = deviceId;
+            this.hubImage = hubImage;
+            this.iotHub = iotHub;
+            this.optimizeForPerformance = optimizeForPerformance;
+            this.proxy = proxy;
+            this.registries = registries;
+        }
+
+        // Returns a DateTime representing the moment the configuration was sent to IoT Hub. Some
+        // of the test modules begin sending events as soon as they are launched, so this timestamp
+        // can be used as a reasonable starting point when listening for events on the IoT hub's
+        // Event Hub-compatible endpoint.
+        public async Task<DateTime> DeployConfigurationAsync(Action<EdgeConfigBuilder> withConfig, CancellationToken token)
+        {
+            var builder = new EdgeConfigBuilder(this.deviceId);
+            builder.AddRegistryCredentials(this.registries);
+            builder.AddEdgeAgent(this.agentImage).WithProxy(this.proxy);
+            builder.AddEdgeHub(this.hubImage, this.optimizeForPerformance)
+                .WithEnvironment(new[] { ("RuntimeLogLevel", "debug") })
+                .WithProxy(this.proxy);
+
+            withConfig(builder);
+
+            DateTime deployTime = DateTime.Now;
+            await builder.Build().DeployAsync(this.iotHub, token);
+
+            return deployTime;
+        }
+
+        public async Task WaitForModulesRunningAsync(EdgeModule[] modules, CancellationToken token)
+        {
+            var agent = new EdgeModule("edgeAgent", this.deviceId);
+            var hub = new EdgeModule("edgeHub", this.deviceId);
+
+            var allModules = new List<EdgeModule> { agent, hub };
```

Comment:
```suggestion
            var allModules = new HashSet<EdgeModule> { agent, hub };
```

Reply:
Suggest to use HashSet to avoid duplication; may need to implement IEquatable<T> for EdgeModule.

Reply:
If the test passes in duplicate module names, that's a bug in the test. We don't want to hide the bug by silently discarding one of the duplicate modules. Here's what happens with the current code:

```
  X TempSensor [726ms]
  Error Message:
   System.ArgumentException : An item with the same key has already been added. Key: tempSensor
  Stack Trace:
     at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at Microsoft.Azure.Devices.Edge.Test.Common.Config.EdgeConfigBuilder.AddModule(String name, String image) in C:\Users\User\Projects\iotedge\test\Microsoft.Azure.Devices.Edge.Test.Common\config\EdgeConfigBuilder.cs:line 51
   at Microsoft.Azure.Devices.Edge.Test.Module.<>c__DisplayClass0_0.<TempSensor>b__1(EdgeConfigBuilder builder) in C:\Users\User\Projects\iotedge\test\Microsoft.Azure.Devices.Edge.Test\Module.cs:line 26
   at Microsoft.Azure.Devices.Edge.Test.Common.Runtime.DeployConfigurationAsync(Action`1 withConfig, CancellationToken token) in C:\Users\User\Projects\iotedge\test\Microsoft.Azure.Devices.Edge.Test.Common\Runtime.cs:line 47
   at Microsoft.Azure.Devices.Edge.Test.Module.TempSensor() in C:\Users\User\Projects\iotedge\test\Microsoft.Azure.Devices.Edge.Test\Module.cs:line 22
   at NUnit.Framework.Internal.AsyncToSyncAdapter.Await(Func`1 invoke) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\AsyncToSyncAdapter.cs:line 99
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(TestExecutionContext context) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 84
```

I think this is the behavior we want.

Reply:
Scratch that...I was looking at duplicates in the deployment; you're talking about duplicates in list of modules to wait for. I'll take a look...

Reply:
Yep, you're right. The expected count of modules will be one more than the actual, so EdgeModule.WaitForStatusAsync will time out. I'll make sure the list is unique.

Reply:
Fixed.